universe-instances-extended/src/Data/Universe/Instances/Extended.hs

```diff
@@ -28,10 +42,46 @@ instance (Representable f, Finite (Rep f), Ord (Rep f), Universe a)
 instance (Representable f, Finite s, Ord s, Finite (Rep f), Ord (Rep f), Universe a)
   => Universe (TracedT s f a)
   where universe = map tabulate universe
+instance (Finite a, Ord a)
+  => Universe (Equivalence a)
+  where universe = [Equivalence ((==) `on` (`M.lookup` m)) | m <- partitions universeF]
```

Comment:
This implementation needs a commentary.

I have hard time convincing myself it's correct, thus I'd error on the fact, that it's probably wrong in some corner case. Help me.

Reply:
Some commentary:

Here's how we enumerate all the equivalence relations for a set X. First we choose an arbitrary element of X. Call it x. Then the equivalence relation can be described in two parts:

1. The equivalence class of x, that is, the set of elements which *is* equivalent to x. (This is required to contain x.)
2. The equivalence relation on all the elements not equivalent to x, which we compute recursively in the same way.

For each piece, we simply nondeterministically choose from among all possibilities. To efficiently track equivalence classes, we choose a fresh number to identify each class, and make a `Map` from elements of X to the identity of their class.

These ideas should shine through pretty clear in the code once you have them in mind:

    -- choose an arbitrary element x of the set X, and set xs = X \ {x}
    go n (x : xs) = do
        -- choose a subset eq of xs to be equivalent to x, and set notEq = xs \ eq
￼        (eq, notEq) <- partitionA (const [False, True]) xs
        -- recursively cook up an equivalence relation on the leftovers, bumping the class identifier up by one
￼        rest        <- go (n + 1) notEq
        -- map x and all the elements of eq to the current class identifier
￼        pure $ M.fromList [(x', n) | x' <- x : eq] `M.union` rest

It is okay to use the boring old list monad instead of our fancy helpers because everything involved here is finite.

Reply:
These comments should be in a implementation.

What's the guarantee that `[Equivalence a]` doesn't contain same elements, i.e its' cardinality is correct. Still have to think too hard.

A test for 3/4/5 element sets would be :+1:, (it turns out that [Bell number](https://en.wikipedia.org/wiki/Bell_number) is a count of partitions of a set)

Also if `partitionA` works so that it returns empty `eq`, will the first element of `universe :: [Equivalence a]` be `Equivalence (==)`. If so, it could be checked as well.

Maybe there should be a big warning on this instance, as Bell numbers grow relatively fast/.

Reply:
> These comments should be in a implementation.

No objection.

> What's the guarantee that `[Equivalence a]` doesn't contain same elements

The usual (strong) inductive one: the call to `partitionA` doesn't return any duplicate equivalence classes for `x`, and by induction there are no duplicate equivalence relations in `rest`. I think this one is not even worth writing down.

Reply:
See https://github.com/dmwit/universe/pull/47/files, adding such doctest would convince me (and readers of documentation) that this works. Using something like `Three` to demonstrate various `Equivalences` would be IMHO neat thing!